support/ebpf/dotnet_tracer.ebpf.c

```diff
@@ -102,12 +102,36 @@ dotnet_find_code_start(PerCPURecord *record, DotnetProcInfo *vi, u64 pc, u64 *co
     // This is unrolled several times, so it needs to be minimal in size.
     // And currently this is the major limit for DOTNET_FRAMES_PER_PROGRAM.
     int orig_pos = pos;
+    if (val == 0) {
+      // pos is fixed to map_elements-2 at this point, and is even.
+      // convert it from u32 to u64 offset
+      int pos64 = (map_elements - 2) / 2 - 1;
+      u64 val64 = scratch->map64[pos64];
 #pragma unroll 256
-    for (int i = 0; i < map_elements - 2; i++) {
-      if (val != 0) {
-        break;
+      // the loop iterations is number of u64 elements minus two:
+      // - the last element special handled earlier
+      // - the second last element which is preloaded immediately above
+      for (int i = 0; i < map_elements / 2 - 2; i++) {
+        if (val64 != 0) {
+          break;
+        }
+        // pos64 will be decremented at most (map_elements - 2) / 2 - 1 times,
+        // so there is no out-of-bounds access.
+        val64 = scratch->map64[--pos64];
+      }
+      // convert u64 offset back to u32 offset
+      pos = pos64 * 2;
+      // use the upper half entry if it is non-zero
+      if (val64 >> 32 != 0) {
+        val64 >>= 32;
+        pos++;
       }
-      val = scratch->map[--pos];
+      val = val64;
+    }
+    // If the pos is less than the offs, it means dirty data was read.
+    if (val == 0 || pos < offs) {
```

Comment:
The `pos` underrun should not happen on valid data as the nibble map should have guaranteed entry for each function. If this happens something else went wrong, and the proper metric is `metricID_UnwindDotnetErrCodeHeader`.

```suggestion
    // If the pos is less than the offs, it means we searched to start of nibble map without finding non-zero entry.
    if (pos < offs) {
      goto bad_code_header;
    }
    if (val == 0) {
```

Reply:
Also the underrun check must happen vefore `val == 0` check